### ISSUE TITLE
Rule: use a particular method of accessing range locations

### ISSUE THREAD

**Issue by not-an-aardvark** [not-an-aardvark/eslint-plugin-eslint-plugin#6]
Nodes in an AST generally have duplicated range information: `start` and `end` properties from acorn, and `range` array properties. This rule would enforce the usage of one or the other.

```js
// (this isn't a very good rule name, there should probably be a better one.)

/* eslint eslint-plugin/use-range-property: ["error", "always"] */

// Invalid:

foo.start;
foo.end;

// Valid:

foo.range[0];
foo.range[1];
```

```js
/* eslint eslint-plugin/use-range-property: ["error", "never"] */

// Invalid:

foo.range[0];
foo.range[1];

// Valid:

foo.start;
foo.end;
```

**Comment by not-an-aardvark** [comment]
With `use-range-property: always`, the rule will need to somehow distinguish between `node.loc.start` and `node.loc.end` (which should always be allowed) and `node.start` and `node.end` (which should be disallowed).

**Comment by not-an-aardvark** [comment]
Closing because it probably isn't possible to disallow `node.start` while allowing `node.loc.start` in the general case.

**Comment by SimenB** [comment]
> Closing because it probably isn't possible to disallow `node.start` while allowing `node.loc.start` in the general case.

"the general" case here refers to if people have destructured or something like that?

If so, could you have as a caveat that there might be false positives and only `loc.start` or `loc.end` is allowed, or something like that?

Not perfect by any means, but still useful (e.g. would have prevented https://github.com/jest-community/eslint-plugin-jest/pull/221 and by extension https://github.com/typescript-eslint/typescript-eslint/issues/175)

**Comment by j-f1** [comment]
This could be implemented as a plugin that parses the code using a configured plugin and adds a getter that throws when accessing `node.start` or `node.end` to all the nodes. This could be used only when running tests for the plugin.
